### PR TITLE
Dual queue mode

### PR DESCRIPTION
**Pull request by odygrd**
Move things around in folders
Add a serialization fast path
Rewrite the SPSC queue
Add DUAL_QUEUE_MODE

### CODE REVIEW

File: quill/test/BoundedSPSCEventQueueTest.cpp [hunk 1]
```diff
@@ -355,4 +354,53 @@ TEST_CASE("produce_consume_many_multithreaded")
   consumer_thread.join();
 }
 
+// TEST_CASE("produce_consume_many_multithreaded_2")
```
Comment (odygrd) [inline quill/test/BoundedSPSCEventQueueTest.cpp:357]: uncomment

File: quill/include/quill/detail/LogMacros.h [hunk 2]
```diff
@@ -54,8 +54,8 @@
   do {                                                                                             \
     static constexpr char const* function_name = __FUNCTION__;                                     \
     struct {                                                                                       \
-      constexpr quill::detail::LogRecordMetadata operator()() const noexcept {                     \
-        return quill::detail::LogRecordMetadata{QUILL_STRINGIFY(__LINE__), __FILE__,               \
+      constexpr quill::LogMacroMetadata operator()() const noexcept {                     \
```
Comment (odygrd) [inline quill/include/quill/detail/LogMacros.h:57]: align \

File: quill/include/quill/detail/LogMacros.h [hunk 3]
```diff
@@ -71,8 +71,8 @@
...
```
Comment (odygrd) [inline quill/include/quill/detail/LogMacros.h:74]: align \

File: quill/include/quill/detail/ThreadContext.h [hunk 4]
```diff
@@ -30,9 +31,11 @@ class ThreadContext
 {
 public:
 #if defined(QUILL_USE_BOUNDED_QUEUE)
-  using SPSCQueueT = BoundedSPSCQueue<RecordBase, QUILL_QUEUE_CAPACITY>;
+  using ObjectSPSCQueueT = BoundedSPSCEventQueue<BaseEvent, QUILL_QUEUE_CAPACITY>;
```
Comment (odygrd) [inline quill/include/quill/detail/ThreadContext.h:34]: rename to event

File: quill/include/quill/detail/ThreadContext.h [hunk 5]
```diff
@@ -30,9 +31,11 @@ class ThreadContext
...
+  using RawSPSCQueueT = BoundedSPSCRawQueue<QUILL_QUEUE_CAPACITY>;
 #else
-  using SPSCQueueT = UnboundedSPSCQueue<BaseEvent>;
+  using ObjectSPSCQueueT = UnboundedSPSCEventQueue<BaseEvent, QUILL_QUEUE_CAPACITY>;
```
Comment (odygrd) [inline quill/include/quill/detail/ThreadContext.h:37]: rename to event

File: quill/include/quill/detail/ThreadContext.h [hunk 6]
```diff
@@ -63,18 +66,41 @@ class ThreadContext
...
   /**
-   * @return A reference to the single-producer-single-consumer queue const overload
+   * In this queue we store only log statements that contain 100% of built-in types
+   * @return A reference to the fast single-producer-single-consumer queue
    */
-  QUILL_NODISCARD_ALWAYS_INLINE_HOT SPSCQueueT const& spsc_queue() const noexcept
+  QUILL_NODISCARD_ALWAYS_INLINE_HOT RawSPSCQueueT& raw_spsc_queue() noexcept
   {
-    return _spsc_queue;
+    return _fast_spsc_queue;
```
Comment (odygrd) [inline quill/include/quill/detail/ThreadContext.h:91]: rename

File: quill/include/quill/Logger.h [hunk 7]
```diff
@@ -105,24 +109,102 @@ class alignas(detail::CACHELINE_SIZE) Logger
...
+    uint64_t timestamp{detail::rdtsc()};
+  #else
+    uint64_t timestamp{static_cast<uint64_t>(std::chrono::system_clock::now().time_since_epoch().count())};
+  #endif
+
+    memcpy(write_buffer, &timestamp, sizeof(timestamp));
+    write_buffer += sizeof(timestamp);
+
+    // Then write the pointer to the LogDataNode. The LogDataNode has all details on how to
+    // deserialize the object. We wil just serialize the arguments in our queue but we need to look
```
Comment (bl-ue) [inline quill/include/quill/Logger.h:157]:
```suggestion
    // deserialize the object. We will just serialize the arguments in our queue but we need to look
```

**Comment by bl-ue** [comment]
Hi @odygrd, I noticed you've added content to the 1.5.3 section to the readme, but you haven't actually released that version.  Are you planning to merge 1.5.3 into 1.6.0?

**Comment by odygrd** [comment]
> Hi @odygrd, I noticed you've added content to the 1.5.3 section to the readme, but you haven't actually released that version. Are you planning to merge 1.5.3 into 1.6.0?

Yes I am thinking to just release `1.6.0` after this PR is merged instead of releasing `1.5.3` before merging and then `1.6.0`. Is there any changes from version `1.5.3` that you might need without the need of upgrading to `1.6.0` ?

**Comment by bl-ue** [comment]
> Is there any changes from version 1.5.3 that you might need without the need of upgrading to 1.6.0 ?

No, I was just wondering if you were going to keep 1.5.3 in the changelog.

> and it looks like a good improvement

👍

**Pull request merged**